crates/router/src/core/user.rs

```diff
@@ -1982,3 +1984,190 @@ pub async fn check_two_factor_auth_status(
+                config.into(),
+                &user_auth_encryption_key,
+            )
+            .await
+            .change_context(UserErrors::InternalServerError)
+            .attach_printable("Failed to encrypt auth config")
+        })
+        .await
+        .transpose()?
+        .map(Into::into);
+
+    let auth_methods = state
+        .store
+        .list_user_authentication_methods_for_owner_id(&req.owner_id)
+        .await
+        .change_context(UserErrors::InternalServerError)
+        .attach_printable("Failed to get list of auth methods for the owner id")?;
+
+    let auth_id = auth_methods
+        .first()
+        .map(|auth_method| auth_method.auth_id.clone())
+        .unwrap_or(uuid::Uuid::new_v4().to_string());
+
+    let now = common_utils::date_time::now();
+    state
+        .store
+        .insert_user_authentication_method(UserAuthenticationMethodNew {
+            id: uuid::Uuid::new_v4().to_string(),
+            auth_id,
+            owner_id: req.owner_id,
+            owner_type: req.owner_type,
+            auth_method: req.auth_method,
+            config: encrypted_config,
+            allow_signup: req.allow_signup,
+            created_at: now,
+            last_modified_at: now,
+        })
+        .await
+        .to_duplicate_response(UserErrors::UserAuthMethodAlreadyExists)?;
+
+    Ok(ApplicationResponse::StatusOk)
+}
+
+pub async fn update_user_authentication_method(
+    state: SessionState,
+    req: user_api::UpdateUserAuthenticationMethodRequest,
+) -> UserResponse<()> {
+    let user_auth_encryption_key = hex::decode(
+        state
+            .conf
+            .user_auth_methods
+            .get_inner()
+            .encryption_key
+            .clone()
+            .expose(),
+    )
+    .change_context(UserErrors::InternalServerError)
+    .attach_printable("Failed to decode DEK")?;
+
+    let auth_config = serde_json::to_value(req.config)
+        .change_context(UserErrors::AuthConfigParsingError)
+        .attach_printable("Failed to convert auth config to json")?;
+
+    let updated_encrypted_config = domain::types::encrypt::<serde_json::Value, masking::WithType>(
+        auth_config.into(),
+        &user_auth_encryption_key,
+    )
+    .await
+    .change_context(UserErrors::InternalServerError)?
+    .into();
+
+    state
+        .store
+        .update_user_authentication_method(
+            &req.id,
+            UserAuthenticationMethodUpdate::UpdateConfig {
+                config: Some(updated_encrypted_config),
+            },
+        )
+        .await
+        .change_context(UserErrors::InvalidUserAuthMethodOperation)?;
+    Ok(ApplicationResponse::StatusOk)
+}
+
+pub async fn list_user_authentication_methods(
+    state: SessionState,
+    req: user_api::GetUserAuthenticationMethodsRequest,
+) -> UserResponse<Vec<user_api::UserAuthenticationMethodResponse>> {
+    let user_auth_encryption_key = hex::decode(
```

Comment:
We should not do KMS decode every time, it will increase the latency and the cost. We can instead cache the decoded data. 

Let's reconsider encrypting the whole config payload : Do we really need to encrypt the whole config data? Because the response needs only non sensitive data from config. By eliminating encryption part in non sensitive data, we can skip this decryption part and also the KMS part in this API.